### ISSUE TITLE
get-disconnectors or get-pendingexportdeletes running incredibly slow on some MIM management agent while fast on another

### ISSUE THREAD

**Issue by PKreyerhoff** [lithnet/miis-powershell#33]
I'm having the issue that on some of my MIM (v4.6.263.0) Management Agents the command get-disconnectors or get-pendingexportdeletes takes ages to return the results while on another one it runs quite fast. 

Examples: 
This one goes fast....  
![image](https://user-images.githubusercontent.com/78021133/105836761-f6e3f100-5fcd-11eb-9c34-da0877cfbde2.png)

while this one runs slow:
![image](https://user-images.githubusercontent.com/78021133/105837225-ab7e1280-5fce-11eb-8765-2bd86d80b359.png)

I chancelled it after 7 minutes and had zero results returned. Same with the get-pendingexportdeletes and also get-csobject is running way slow on that Management Agent. On the other it is again fast and I don't have any clue why though? Both Agents are for ADservices and it's connector space is pointing to the same sql instance. 

Did you ever face such an issue? 

I have some unwanted export deletes outstanding which need to be removed and it took now more than 4 hours to delete 158 objects for only one agent.

**Comment by ryannewington** [comment]
@PKreyerhoff 

If you perform these actions in the MIM GUI do you see the same results? The PowerShell cmdlets call the same APIs as the UI, so that can help us narrow down the problem

99% of the time that I have seen this sort of behavior is because indexes and/or statistics in the FIMSync database being out of whack. Try and index rebuild and update stats and see if the problem goes away.

**Comment by PKreyerhoff** [comment]
@ryannewington 

Within the MIM GUI the behaviour is different. I can show the disconnectors for any agent quite fast. If I search for the explicit disconnectors it loads within seconds. 

Anyway I'll give the indexes a try and let you know.

**Comment by ryannewington** [comment]
We might be making a slightly different call than the GUI, but this could still be enough to trigger a different query path in SQL. Try the index rebuild and `exec sp_updatestats` and see if that fixes.

**Comment by PKreyerhoff** [comment]
@ryannewington 

No joy with the indexes. Though slightly. Some stats from the joiner.log on the command "Get-Disconnectors -MA xxxx -Type Explicit | Set-DisconnectorState -Type Normal" I'm running currently to have these deletions removed. 

This was the one that run fast: (ms per event)
![image](https://user-images.githubusercontent.com/78021133/105996652-ef464a00-60aa-11eb-8bb6-49189b990507.png)

this was very slow: (1-2 mins per event)
![image](https://user-images.githubusercontent.com/78021133/105996866-32082200-60ab-11eb-9920-a09e07cbf8d0.png)

now since we rebuild the indexes: (40 secs per event)
![image](https://user-images.githubusercontent.com/78021133/105997055-70054600-60ab-11eb-8e41-d2d387d16635.png)

Any other ideas which we can try to get more thoughput?

**Comment by ryannewington** [comment]
Hmmmm I'm not too sure. My next step would be to perform an SQL profiler trace, to see where the bottleneck is. It's unlikely to be the code itself, since some MAs perform fast and others slow. I'm certain it's DB related, but it gets complicated at this point. My process is that I look at expensive queries in profiler, and if necessary analyse the query plans for those queries and see if the problem presents itself... It's really getting into the weeds of SQL though.

There could be an issue with temp db config or performance as well.

Sorry it's a bit of guess work from my end.

**Comment by PKreyerhoff** [comment]
@ryannewington 

We found the error or let's say the issue which is maybe due to our agents setup. With ProcMon I found a RPC connection to the Domain Controller which is set in the AD Management Agent which was blocked by the Firewall. 

![image](https://user-images.githubusercontent.com/78021133/106142514-dd7baa00-6171-11eb-86c8-a1cb3884e90d.png)

Since we opened the port 135 to the DCs of the Mangement Agents we don't have any issues. 
I don't get though why it connects to the DC? I thought it is only searching in the connector space?
